DOCSP-40787 - Connection Options (#39)
mongodb/docs-cpp: source/connect/connection-options.txt

.. _cpp-connection-options:

==========================
Specify Connection Options
==========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 1
   :class: singlecol

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: connection string, URI, server, Atlas, settings, configure

Overview
--------

This page describes the MongoDB connection and authentication options

> available in the {+driver-short+}. You can configure your connection by using
> options in the connection URI.

Reviewer: S: since you can use the connection URI or a mongocxx::options::client instance:
```suggestion
available in the {+driver-short+}. You can configure your connection by specifying
options in the connection URI or in an instance of the ``mongocxx::options::client`` class.
```

.. note::

   You can set most connection options by using either the connection URI or the

>    ``client_options`` parameter. However, some options support only one of these

Reviewer: S: I think it's helpful to specify that this parameter should be an instance of mongocxx::options::client
```suggestion
  a ``mongocxx::options::client`` instance as a parameter. However, some options support only one of these
```

   methods. You might need to use both a connection URI and the ``client_options``
   parameter to set all the options that you need.

.. _cpp-connection-uri:

Using the Connection URI
------------------------

When you construct a ``mongocxx::client`` object, you can pass in a ``mongocxx::uri``
object that represents your connection URI. This connection URI can include
connection options as ``<name>=<value>`` pairs. In the following example,
the connection URI contains the ``tls`` option with a value of ``true`` and the
``tlsCertificateKeyFile`` option with a value of ``path/to/file.pem``:

.. literalinclude:: /includes/connect/connection-options.cpp
   :language: cpp
   :copyable: true

> 

Reviewer: S: you can add a section here (maybe called "Using an Options Parameter") about how to specify options by setting fields of a `mongocxx::client::options` object, and passing that object to the `mongocxx:client` constructor

Connection URI Options
----------------------

> 

Reviewer: I: when I click the server manual table links, they all just direct me to the top of the "Connection Strings" page rather than different sections

Author: Unfortunately, no. [I asked DOP](https://mongodb.slack.com/archives/C018UTGP23T/p1719620843807509) and they've opened a ticket for it. In the meantime, I updated the language to not say it links directly to that option on the page.

The following sections describe the connection options that you can set in the connection
URI passed to the {+driver-short+}. Each connection option links to its entry in
the {+mdb-server+} manual. If you can access the value of a connection option by using

> a member of the ``mongocxx::uri`` class, the member name is also listed. 

Reviewer: Q: can you clarify what a member of the mongocxx::uri class means and how that's different than using name=value pairs? And maybe add a code example that uses a mongocxx::uri member function to the "Using the Connection URI" section?

Replica Set Options
~~~~~~~~~~~~~~~~~~~

.. list-table::
   :header-rows: 1
   :stub-columns: 1

   * - Connection URI Option
     - ``mongocxx::uri`` Member

   * - :manual:`directConnection </reference/connection-string/#mongodb-urioption-urioption.directConnection>`
     - `direct_connection() <https://mongocxx.org/api/current/classmongocxx_1_1v__noabi_1_1uri.html#a4d4e43c3ef3f30f81c990c04082b1655>`__
   
   * - :manual:`replicaSet </reference/connection-string/#mongodb-urioption-urioption.replicaSet>`
     - `replica_set() <https://mongocxx.org/api/current/classmongocxx_1_1v__noabi_1_1uri.html#ad2d71ae778881db9ffc64fc7c07e0dad>`__
   
Connection Options
~~~~~~~~~~~~~~~~~~

TLS Options
```````````

.. list-table::
   :header-rows: 1
   :stub-columns: 1

   * - Connection URI Option
     - ``mongocxx::uri`` Member

> 

Reviewer: S: I think it makes sense to add a "mongocxx::client::options Member" row - especially to this section's table, since seems like most of the options you can set in an options parameter are tls / ssl related (from [API docs](https://mongocxx.org/api/current/classmongocxx_1_1v__noabi_1_1options_1_1client.html)).
```suggestion
- ``mongocxx::client::options`` Member
```

Author: LMK if other page changes are sufficient here.

.. _cpp-connection-options:

==========================
Specify Connection Options
==========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 1
   :class: singlecol

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: connection string, URI, server, Atlas, settings, configure

Overview
--------

This page describes the MongoDB connection and authentication options
available in the {+driver-short+}. You can configure your connection by
specifying options in the connection URI or by passing an instance of the
``mongocxx::options::client`` class as the ``client_options`` parameter to the
``mongocxx::client`` constructor.

.. note::

   You can specify some connection options only in the connection URI, and others only
   in an the ``client_options`` parameter.
   You might need to combine these methods to specify all the options that you need.

> 

Reviewer: S: might be helpful to put the next two sections ("Using the Connection URI" and "Using a ``mongocxx::options::client`` Object") in their own section (called something like "Set Connection Options"), rather than in "Overview." That way, the top-level headers seem a little more complete (Overview, Set Options, Read Options, Connection URI Table)
```suggestion
Set Connection Options
----------------------
```
